templates/components/multiagent/typescript/workflow/tools.ts

```diff
@@ -0,0 +1,324 @@
+  }
+  return new QueryEngineTool({
+    queryEngine: (index as any).asQueryEngine({
+      similarityTopK: topK,
+    }),
+    metadata: {
+      name: "retriever",
+      description: `Use this tool to retrieve information about the provided data.`,
+    },
+  });
+};
+
+export const getAvailableTools = async () => {
+  const configFile = path.join("config", "tools.json");
+  let toolConfig: any;
+  const tools: BaseToolWithCall[] = [];
+  try {
+    toolConfig = JSON.parse(await fs.readFile(configFile, "utf8"));
+  } catch (e) {
+    console.info(`Could not read ${configFile} file. Using no tools.`);
+  }
+  if (toolConfig) {
+    tools.push(...(await createTools(toolConfig)));
+  }
+  const queryEngineTool = await getQueryEngineTool();
+  if (queryEngineTool) {
+    tools.push(queryEngineTool);
+  }
+
+  return tools;
+};
+
+export const lookupTools = async (
+  toolNames: string[],
+): Promise<BaseToolWithCall[]> => {
+  const availableTools = await getAvailableTools();
+  return availableTools.filter((tool) =>
+    toolNames.includes(tool.metadata.name),
+  );
+};
+
+/**
+ * Call multiple tools and return the tool messages
+ */
+export const callTools = async (
+  toolCalls: ToolCall[],
+  tools: BaseToolWithCall[],
+  ctx: Context,
+  agentName: string,
+  writeEvent: boolean = true,
+  // eslint-disable-next-line max-params
+): Promise<ChatMessage[]> => {
+  const toolMsgs: ChatMessage[] = [];
+  if (toolCalls.length === 0) {
+    return toolMsgs;
+  }
+  if (toolCalls.length === 1) {
+    const tool = tools.find((tool) => tool.metadata.name === toolCalls[0].name);
+    if (!tool) {
+      throw new Error(`Tool ${toolCalls[0].name} not found`);
+    }
+    return [
+      (await callSingleTool(toolCalls[0], tool, (msg: string) => {
+        if (writeEvent) {
+          ctx.writeEventToStream(
+            new AgentRunEvent({
+              name: agentName,
+              text: msg,
+              type: "text",
+            }),
+          );
+        }
+      })) as ChatMessage,
+    ];
+  }
+  // Multiple tool calls, show events in progress
+  const progressId = crypto.randomUUID();
+  const totalSteps = toolCalls.length;
+  let currentStep = 0;
+  for (const toolCall of toolCalls) {
+    const tool = tools.find((tool) => tool.metadata.name === toolCall.name);
+    if (!tool) {
+      throw new Error(`Tool ${toolCall.name} not found`);
+    }
+    const toolMsg = await callSingleTool(toolCall, tool, (msg: string) => {
+      ctx.writeEventToStream(
+        new AgentRunEvent({
+          name: agentName,
+          text: msg,
+          type: "progress",
+          data: {
+            id: progressId,
+            total: totalSteps,
+            current: currentStep,
+          },
+        }),
+      );
+      currentStep++;
+    });
+    toolMsgs.push(toolMsg as ChatMessage);
+  }
+  return toolMsgs;
+};
+
+export const callSingleTool = async (
+  toolCall: ToolCall,
+  tool: BaseToolWithCall,
+  eventEmitter: (msg: string) => void,
```

Comment:
eventEmitter should be optional (you're actually checking below)
```suggestion
  eventEmitter?: (msg: string) => void,
```